contracts/DEXes/UniV3StablesJoint.sol

```diff
@@ -0,0 +1,639 @@
+            .getAmountsForLiquidity(
+                _slot0.sqrtPriceX96,
+                TickMath.getSqrtRatioAtTick(minTick),
+                TickMath.getSqrtRatioAtTick(maxTick),
+                positionInfo.liquidity
+            );
+        // uniswap orders token0 and token1 based on alphabetical order
+        return tokenA < tokenB ? (amount0, amount1) : (amount1, amount0);
+    }
+
+    /*
+     * @notice
+     *  Function returning the amount of rewards earned until now - unclaimed
+     * @return uint256 array of tokenA and tokenB earned as rewards
+     */
+    function pendingRewards() public view override returns (uint256[] memory) {
+        // Initialize the array to same length as reward tokens
+        uint256[] memory _amountPending = new uint256[](rewardTokens.length);
+
+        // Get LP position info
+        IUniswapV3Pool.PositionInfo memory positionInfo = _positionInfo();
+
+        // Initialize to the current status of owed tokens
+        (_amountPending[0], _amountPending[1]) = tokenA < tokenB
+            ? (positionInfo.tokensOwed0, positionInfo.tokensOwed1)
+            : (positionInfo.tokensOwed1, positionInfo.tokensOwed0);
+
+        // Gas savings
+        IUniswapV3Pool _pool = IUniswapV3Pool(pool);
+        int24 _minTick = minTick;
+        int24 _maxTick = maxTick;
+
+        // Use Uniswap views library to calculate the fees earned in tokenA and tokenB based
+        // on current status of the pool and provided position
+        (uint128 tokensOwed0, uint128 tokensOwed1) = UniswapHelperViews.getFeesEarned(
+            UniswapHelperViews.feesEarnedParams(
+                positionInfo.liquidity,
+                _pool.slot0().tick,
+                _minTick,
+                _maxTick,
+                _pool.feeGrowthGlobal0X128(),
+                _pool.feeGrowthGlobal1X128(),
+                positionInfo.feeGrowthInside0LastX128,
+                positionInfo.feeGrowthInside1LastX128
+            ),
+            _pool.ticks(_minTick),
+            _pool.ticks(_maxTick)
+        );
+
+        // Reorder to make sure amounts are added correctly
+        if (tokenA < tokenB) {
+            _amountPending[0] += tokensOwed0;
+            _amountPending[1] += tokensOwed1;
+        } else {
+            _amountPending[1] += tokensOwed0;
+            _amountPending[0] += tokensOwed1;
+        }
+
+        return _amountPending;
+    }
+
+    /*
+     * @notice
+     *  Function called by the uniswap pool when minting the LP position (providing liquidity),
+     * instead of approving and sending the tokens, uniV3 calls the callback imoplementation
+     * on the caller contract
+     * @param amount0Owed, amount of token0 to send
+     * @param amount1Owed, amount of token1 to send
+     * @param data, additional calldata
+     */
+    function uniswapV3MintCallback(
+        uint256 amount0Owed,
+        uint256 amount1Owed,
+        bytes calldata data
```

Comment:
just in case should we require data to be empty?

Reply:
to avoid someone sending us a malicious payload...

but the msg.sender require should be enough